src/openlrr/game/world/Roof.cpp

```diff
@@ -22,30 +24,304 @@ LegoRR::Roof_Globs & LegoRR::roofGlobs = *(LegoRR::Roof_Globs*)0x00553180;
+		vertices[0].position.x = vertPos3->x;
+		vertices[0].position.y = vertPos3->y;
+		vertices[0].position.z = vertPos3->z;
+		/* &= ~ROOF_SHIFTVERTICES; // (~0x4) */
+		locRoofBlock->flags &= ~ROOF_FLAG_SHIFTVERTICES;
+	}
+	// Finish clearing normals
+	vertices[3].normal.x = 0.0;
+	vertices[3].normal.z = 0.0;
+	vertices[2].normal.x = 0.0;
+	vertices[2].normal.z = 0.0;
+	vertices[1].normal.x = 0.0;
+	vertices[1].normal.z = 0.0;
+	vertices[0].normal.x = 0.0;
+	vertices[0].normal.z = 0.0;
+	vertices[3].normal.y = 0.0;
+	vertices[2].normal.y = 0.0;
+	vertices[1].normal.y = 0.0;
+	vertices[0].normal.y = 0.0;
+	/* white (no coloration) */
+	vertices[3].colour = 0xffffffff;
+	vertices[2].colour = 0xffffffff;
+	vertices[1].colour = 0xffffffff;
+	vertices[0].colour = 0xffffffff;
+	vertices[3].tu = 0.0;
+	vertices[3].tv = 1.0;
+	vertices[2].tu = 1.0;
+	vertices[2].tv = 1.0;
+	vertices[1].tu = 1.0;
+	vertices[1].tv = 0.0;
+	vertices[0].tu = 0.0;
+	vertices[0].tv = 0.0;
+	Gods98::Container_Mesh_SetVertices(roofGlobs.contMesh, locRoofBlock->groupID, 0, 4, vertices);
+}
 
 // <LegoRR.exe @004634f0>
-//void __cdecl LegoRR::Roof_Update(void);
+void __cdecl LegoRR::Roof_Update(void)
+{
+	RoofBlock* locRoofBlock;
+	float fVar2;
+	uint32 dirIndex;
+	uint32 width;
+	Point2I* pDir;
+	Vector3F worldPos;
+	Vertex vertices[4];
+
+	// Check if the roof needs an update, then go through loop, updating object vectors
+	if (roofGlobs.needsUpdate != 0) {
+		roofGlobs.needsUpdate = 0;
+		width = roofGlobs.dimensions.width;
+		// Sanity check then iterate through heigh and with
+		if (roofGlobs.dimensions.height != 0) {
+			for (uint32 y = 0; y < roofGlobs.dimensions.height; y++) {
+				if (width != 0) {
+					for(uint32 x = 0; x < width; x++) {
+						/* Roof vertex has been updated by RoofMesh_LowerRoofVertices.
+							Update it here, and unset the flag. */
+						locRoofBlock = roofGlobs.grid + (y * width + x);
+						if ((roofGlobs.grid[(y * width + x)].flags & ROOF_FLAG_NEEDUPDATE) != 0) {
+							// Setup all directions with appropritate vectors		
+							Point2I DIRECTIONS_4X4[16] = { {0,0},{-1,0},{0,1},{-1,1},{0,0},{1,0},{1,1},{0,1},{0,0},{0,-1},{1,-1},{1,0},{0,0},{-1,-1},{0,-1},{-1,0}};
+							// Grab verticies to manipulate below
+							Gods98::Container_Mesh_GetVertices(roofGlobs.contMesh, locRoofBlock->groupID, 0, 4, vertices);							
+							/* loop for DIRECTIONS_4X4[i...][] */
+							for(uint32 i = 0; i < 4; i++) {
+								/* Some counter-clockwise flag maybe? */
+								dirIndex = i;
+								/* GHIDRA FAIL: setting 0x4 flag equate causes Ghidra to hang
+								   if (roof->flags & ROOF_SHIFTVERTICES) // (0x4) */
+								if ((locRoofBlock->flags & ROOF_FLAG_SHIFTVERTICES) != 0) {
+									dirIndex = i - 1 & 3;
+								}
+								/* loop for DIRECTIONS_4X4[dirIndex][j...] */
+								for(int j =4; j != 0; j--){
+									if (Roof_GetBlockPlaneNormal(x + DIRECTIONS_4X4[dirIndex * 4].x, DIRECTIONS_4X4[dirIndex * 4].y + y, &worldPos) != 0) {
```

Comment:
`DIRECTIONS_4X4` also has a `j` lookup component, so `DIRECTIONS_4X4[dirIndex * 4 + j]`. You can also simplify this to a 2D array if you want (`DIRECTIONS_4X4[dirIndex][j]`, where every 4 elements would be split up into a sub-array).

```cpp
Point2I DIRECTIONS_4X4[4][4] = {
    { {0,0}, {-1, 0}, {0, 1}, {-1,1} },
    { {0,0}, { 1, 0}, {1, 1}, { 0,1} },
    { {0,0}, { 0,-1}, {1,-1}, { 1,0} },
    { {0,0}, {-1,-1}, {0,-1}, {-1,0} }
};
```

Because of this, the for loop actually shouldn't be reversed. It's only reversed in Ghidra because the `j` variable isn't used directly, instead a pointer to `DIRECTIONS_4X4` was incremented each loop.